code/network/lib/Loot/Network/Message.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE DataKinds                 #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE KindSignatures            #-}
+{-# LANGUAGE TypeFamilies              #-}
+{-# LANGUAGE TypeFamilyDependencies    #-}
+
+-- | Networking utilities. Playground for now.
+
+module Loot.Network.Message
+       ( Message (..)
+       , Callback (..)
+       , CallbackWrapper (..)
+       , handler
+       , createDMap
+       , runCallback
+       , runCallbackInt
+       ) where
+
+import Data.Dependent.Map (DSum ((:=>)))
+import qualified Data.Dependent.Map as D
+import Data.GADT.Compare ((:~:) (Refl), GEq (..), GOrdering (..))
+import Data.Reflection (reifyNat)
+import Data.Singletons.TypeLits hiding (natVal)
+import Unsafe.Coerce (unsafeCoerce)
+
+class (KnownNat (MsgTag d), Binary d) => Message d where
+    type family MsgTag d = (n :: Nat) | n -> d
+
+-- | Action called on a particular type.
+data Callback m a n =
+    forall d. (Message d, MsgTag d ~ n) =>
+    Callback { unCallback :: d -> m a }
+
+-- | Packed callback.
+data CallbackWrapper m a =
+    forall n. KnownNat n => CallbackWrapper { unCallbackWrapper :: Callback m a n }
+
+-- | Convenient helper to create callback wrappers.
+handler :: Message d => (d -> m a) -> CallbackWrapper m a
+handler = CallbackWrapper . Callback
+
+-- Is there a better way to do this w/o using unsafeCoerce?
+instance GEq (Sing :: Nat -> *) where
+    geq (SNat :: SNat n) (SNat :: SNat m) =
+        if natVal (Proxy @n) /= natVal (Proxy @m)
+        then Nothing
+        else Just (unsafeCoerce Refl)
+
+instance D.GCompare (Sing :: Nat -> *) where
+    gcompare (SNat :: SNat n) (SNat :: SNat m) =
+        case compare (natVal (Proxy @n)) (natVal (Proxy @m)) of
+          LT -> GLT
+          EQ -> unsafeCoerce GEQ
+          GT -> GGT
+
+-- | Creates a dmap from callbacks. If more than one callback with the
+-- same incoming message typewill be in the list, only first will be
+-- put in.
+createDMap :: forall m a. [CallbackWrapper m a] -> D.DMap Sing (Callback m a)
+createDMap callbacks =
+    D.fromList $
+    map (\(CallbackWrapper c@(_ :: Callback m a n)) -> ((SNat :: SNat n) :=> c))
+        callbacks
+
+-- | Executes a callback given a dmap, switching integer and BS.
+runCallback :: D.DMap Sing (Callback m a) -> SNat n -> ByteString -> m a
+runCallback dmap sn bs = case dmap D.! sn of (Callback x) -> x (parse bs)
+
+-- | Same as 'runCallback', but accepts value-level integer.
+runCallbackInt :: D.DMap Sing (Callback m a) -> Integer -> ByteString -> m a
+runCallbackInt dmap i bs =
+    reifyNat i $ \(Proxy :: Proxy n) -> runCallback dmap (SNat :: SNat n) bs
```

Comment:
1-2: these questions don't really make sense, but it will probably fail and loop in `reifyNat`. In the third case, this will result in a runtime exception. But my idea is that subsystem only claims the ability to handle certain list of message tags and networking won't pass any other to it. If it does, it's a bug worth `error`ing.

Reply:
If so, then its ok.